src/main/java/eu/europeana/metis/sandbox/service/workflow/HarvestServiceImpl.java

```diff
@@ -96,15 +104,8 @@ public void harvestOaiPmh(String datasetId, Record.RecordBuilder recordDataEncap
           return ReportingIteration.IterationResult.TERMINATE;
         }
 
-        if (datasetService.isXsltPresent(datasetId)) {
-          recordPublishService.publishToTransformationToEdmExternalQueue(
-              harvestOaiRecordHeader(datasetId, completeOaiHarvestData,
-                  recordDataEncapsulated), Step.HARVEST_OAI_PMH);
-        } else {
-          recordPublishService.publishToHarvestQueue(
-              harvestOaiRecordHeader(datasetId, completeOaiHarvestData,
-                  recordDataEncapsulated), Step.HARVEST_OAI_PMH);
-        }
+        recordInfoList.add(harvestOaiRecordHeader(datasetId, completeOaiHarvestData,
```

Comment:
I'm wondering if we could have this return the total records  to the user faster?
That is because the harvesting of the actual records, could take a longer time.
With this call it's actually using the header to harvest the record(The method name is a little missleading perhaps).
Could we get first all the identifiers, report the total, and then start harvesting the records(while setting the status to in progress?).
Not sure if it's possible with the current implementation though..

Reply:
how it behaves now after the change. Is as follows.
![metis-sandbox dataset status states](https://user-images.githubusercontent.com/19560266/159490653-3a49516e-aa2f-41c8-9825-307e684bc388.png)

in order to get the total amount of records first gathers information about records and then they are processed.

the method was renamed to ```harvestOaiRecords```

I think at this moment getting the amount of identifiers or going through the zip file are the expensive operations.

Sequence diagram rendered by https://www.websequencediagrams.com/, with source code:
~~~
title metis-sandbox dataset status states

Browser->Swagger: POST /dataset/{name}/harvestByXXX

note right of Swagger
    User posts dataset
    JSON with new id is returned.
    Harvesting Identifiers
    on the background starts.
end note
Swagger->Browser: 202 Accepted, dataset {id: 1}
Browser->Swagger: GET /dataset/1
Swagger->Browser: 200 OK, json
note right of Swagger
{
  "status": "harvesting identifiers",
  "portal-publish": "Harvesting dataset identifiers and records.",
  "total-records": 0,
  "processed-records": 0,
  "progress-by-step": [],
  "dataset-info": {
    "dataset-id": "2",
    "dataset-name": "test_22032022_1412",
    "creation-date": "2022-03-22T14:12:07.953532",
    "language": "Dutch",
    "country": "Netherlands",
    "record-limit-exceeded": false,
    "transformed-to-edm-external": false
  }
}
end note
Browser->Swagger: GET /dataset/1
Swagger->Browser: 200 OK, json
note right of Swagger
{
  "status": "in progress",
  "portal-publish": "A review URL will be generated when the dataset has finished processing.",
  "total-records": 9,
  "processed-records": 0,
  "progress-by-step": [],
  ...
}
end note
Browser->Swagger: GET /dataset/1
Swagger->Browser: 200 OK, json
note right of Swagger
{
  "status": "completed",
  "portal-publish": "https://metis-sandbox.eanadev.org/portal/search?view=grid&q=edm_datasetName:1_test_22032022_1412*",
  "total-records": 9,
  "processed-records": 9,
  "progress-by-step": [],
  ...
}
end note 

~~~

Reply:
and answering your question, with the current implementation, we need to decouple somehow the harvesting in two steps. But needs redesign I think.